pkg/resource/role/hooks.go

```diff
@@ -185,6 +189,155 @@ func (rm *resourceManager) removePolicy(
+		rlog.Debug(
+			"adding inline policy to role",
+			"policy_name", polName,
+		)
+		err = rm.addInlinePolicy(ctx, desired, polName, polDoc)
+		if err != nil {
+			return err
+		}
+	}
+	for _, pair := range toDelete {
+		polName := pair.Key
+		rlog.Debug(
+			"removing inline policy from role",
+			"policy_name", polName,
+		)
+		if err = rm.removeInlinePolicy(ctx, desired, polName); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getInlinePolicies returns a map of inline policy name and policy docs
+// currently attached to the Role.
+//
+// NOTE(jaypipes): There's no way around the inefficiencies of this method
+// without caching stuff, and I don't think it's useful to have an unbounded
+// cache for these inline policy documents :( IAM's ListRolePolicies API call
+// only returns the *policy names* of inline policies. You need to call
+// GetRolePolicy API call for each inline policy name in order to get the
+// policy document. Yes, they force an O(N) time complexity for this
+// operation...
+func (rm *resourceManager) getInlinePolicies(
+	ctx context.Context,
+	r *resource,
+) (map[string]*string, error) {
+	var err error
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.getInlinePolicies")
+	defer func() { exit(err) }()
+
+	roleName := r.ko.Spec.Name
+
+	input := &svcsdk.ListRolePoliciesInput{}
+	input.RoleName = roleName
+	res := map[string]*string{}
+
+	err = rm.sdkapi.ListRolePoliciesPagesWithContext(
+		ctx, input, func(page *svcsdk.ListRolePoliciesOutput, _ bool) bool {
+			if page == nil {
+				return true
+			}
+			for _, p := range page.PolicyNames {
+				res[*p] = nil
+			}
+			return page.IsTruncated != nil && *page.IsTruncated
+		},
+	)
+	rm.metrics.RecordAPICall("READ_MANY", "ListRolePolicies", err)
+
+	// Now we need to grab the policy documents for each policy name
+	for polName, _ := range res {
+		input := &svcsdk.GetRolePolicyInput{}
+		input.RoleName = roleName
+		input.PolicyName = &polName
+		resp, err := rm.sdkapi.GetRolePolicyWithContext(ctx, input)
+		rm.metrics.RecordAPICall("READ_ONE", "GetRolePolicy", err)
+		if err != nil {
+			return nil, err
+		}
+		cleanedDoc, err := decodeDocument(*resp.PolicyDocument)
+		if err != nil {
+			return nil, err
+		}
+		res[polName] = &cleanedDoc
+
+	}
+	return res, nil
+}
+
+// addInlinePolicy adds the supplied inline Policy to the supplied Role
+// resource
+func (rm *resourceManager) addInlinePolicy(
+	ctx context.Context,
+	r *resource,
+	policyName string,
+	policyDoc *string,
+) (err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.addInlinePolicy")
+	defer func() { exit(err) }()
+
+	input := &svcsdk.PutRolePolicyInput{}
+	input.RoleName = r.ko.Spec.Name
+	input.PolicyName = &policyName
+	cleanedDoc, err := decodeDocument(*policyDoc)
+	if err != nil {
+		return err
+	}
+	input.PolicyDocument = &cleanedDoc
+	_, err = rm.sdkapi.PutRolePolicyWithContext(ctx, input)
+	rm.metrics.RecordAPICall("CREATE", "PutRolePolicy", err)
```

Comment:
```suggestion
rm.metrics.RecordAPICall("UPDATE", "PutRolePolicy", err)
```

Reply:
I think that for inline policies, since they are an attribute of only the single role they exist within, it should be considered UPDATE to add or remove them

Reply:
Agreed. Also changed the Delete/DetachRolePolicy calls from DELETE to UPDATE.